src/storage/src/render/persist_sink.rs

```diff
@@ -89,45 +90,81 @@ pub fn render<G>(
-                        } else {
-                            empty.iter()
-                        }
-                    })
-                    .map(|&(ref row, ref ts, ref diff)| ((row, &()), ts, diff));
+                // See if any timestamps are done!
+                let finalized_timestamps: Vec<_> = stashed_batches
+                    .keys()
+                    .filter(|ts| !input_frontier.less_equal(ts))
+                    .copied()
+                    .collect();
 
                 if PartialOrder::less_than(&*shared_frontier.borrow(), &input_frontier) {
                     // We always append, even in case we don't have any updates, because appending
                     // also advances the frontier.
-                    // TODO(aljoscha): Figure out how errors from this should be reported.
-                    let expected_upper = shared_frontier.borrow().clone();
-                    write
-                        .compare_and_append(updates, expected_upper, input_frontier.clone())
-                        .await
-                        .expect("cannot append updates")
-                        .expect("cannot append updates")
-                        .expect("invalid/outdated upper");
+                    if finalized_timestamps.is_empty() {
+                        let expected_upper = shared_frontier.borrow().clone();
+                        write
+                            .compare_and_append(
+                                Vec::<((SourceData, ()), Timestamp, Diff)>::new(),
+                                expected_upper,
+                                input_frontier.clone(),
+                            )
+                            .await
+                            .expect("cannot append updates")
+                            .expect("cannot append updates")
+                            .expect("invalid/outdated upper");
+
+                        // advance our stashed frontier
+                        *shared_frontier.borrow_mut() = input_frontier.clone();
+                        // wait for more data or a new input frontier
+                        continue;
+                    }
 
+                    for ts in finalized_timestamps {
+                        // TODO(aljoscha): Figure out how errors from this should be reported.
+
+                        let expected_upper = shared_frontier.borrow().clone();
+
+                        let mut batch = stashed_batches
+                            .remove(&ts)
+                            .expect("batch for timestamp to still be there")
+                            .finish(input_frontier.clone())
+                            .await
+                            .expect("invalid usage");
+
+                        write
+                            .compare_and_append_batch(
+                                &mut batch,
+                                expected_upper,
+                                input_frontier.clone(),
```

Comment:
oh i think these are wrong, we need to be able to compare and append multiple things